src/Common/tests/Tests/System/Net/VirtualNetworkStreamTest.cs

```diff
@@ -0,0 +1,145 @@
+                    uint readChecksum = Fletcher32.Checksum(readFrame, 0, readFrame.Length);
+
+                    Assert.Equal(writeChecksum, readChecksum);
+                }
+            }
+        }
+
+        [Fact]
+        public void VirtualNetworkStream_WriteMoreThanRead_Ok()
+        {
+            var network = new VirtualNetwork();
+
+            using (var client = new VirtualNetworkStream(network, isServer: false))
+            using (var server = new VirtualNetworkStream(network, isServer: true))
+            {
+                var writeFrame = new byte[] {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
+                int expectedReceivedBytes = writeFrame.Length / 2;
+
+                client.Write(writeFrame, 0, expectedReceivedBytes);
+
+                var readFrame = new byte [expectedReceivedBytes];
+                server.Read(readFrame, 0, expectedReceivedBytes);
+
+                var expectedFrame = new byte[expectedReceivedBytes];
+                Array.Copy(writeFrame, expectedFrame, expectedReceivedBytes);
+
+                Assert.Equal(expectedFrame, readFrame);
+            }
+        }
+
+        [Fact]
+        public void VirtualNetworkStream_ReadMoreThanWrite_Ok()
+        {
+            var network = new VirtualNetwork();
+
+            using (var client = new VirtualNetworkStream(network, isServer: false))
+            using (var server = new VirtualNetworkStream(network, isServer: true))
+            {
+                var writeFrame = new byte[] { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 };
+                int frame1Len = writeFrame.Length / 2;
+
+                client.Write(writeFrame, 0, frame1Len);
+                client.Write(writeFrame, frame1Len, writeFrame.Length - frame1Len);
+
+                int expectedReceivedBytes = writeFrame.Length;
+                var readFrame = new byte[expectedReceivedBytes];
+
+                int bytesRead = 0;
+
+                do
+                {
+                    bytesRead += server.Read(readFrame, bytesRead, expectedReceivedBytes - bytesRead);
+                }
+                while (bytesRead < expectedReceivedBytes); 
+
+                Assert.Equal(writeFrame, readFrame);
+            }
+        }
+
+        [Fact]
+        public void VirtualNetworkStream_MultiThreadIntegrityTest_Ok()
+        {
+            int maxFrameSize = 2048;
+            Assert.True(maxFrameSize > sizeof(int) + 1);
+
+            var rnd = new Random();
+            var network = new VirtualNetwork();
+            var checksumAndLengths = new ConcurrentDictionary<int, Tuple<uint, int>>();
+
+            object readLock = new object();
+
+            using (var client = new VirtualNetworkStream(network, isServer: false))
+            using (var server = new VirtualNetworkStream(network, isServer: true))
+            {
+                Parallel.For(0, 100000, async (int i) =>
+                {
+                    int bufferSize = rnd.Next(sizeof(int) + 1, maxFrameSize);
```

Comment:
Random isn't thread-safe, so this call should be protected.  Random has a nasty habit when it's used concurrently of getting into a corrupted state where it then only ever produces zeros.